### PR TITLE
Frequency Analysis Widget UI Redesign

### PR DESCRIPTION
**Pull request by Hamms**
Significantly redesigns the Frequency Analysis Widget.
#### Before

image
#### After

image

Major changes of note:
- Removed input box, replaced with dropdown + dialog (similar to text compression widget)
- Rather than swapping substitutions, you now have a bank of unassigned letters you can drag up and assign
- Controls are all isolated into a tabular area, you can now sort message frequency letters and substitution letters independently

### CODE REVIEW

**Comment by joshlory** [comment]
Initial impressions: this is way easier to understand when solving "Sample Message (hard)"!  Love the dragging of the blue final letters instead of the orange starting letters.

A few nits:

1. Can we bring back the locks?  I miss being able to lock letters I'm sure of while I experiment with other substitutions.

2. "Random" here is a little confusing.  Elijah had to explain that this means "Random Cipher".  "Sort" was a lot clearer as a user.

screen shot 2015-09-16 at 12 21 50 pm

3. Prefer "Sort originals" above "Sort substitutions" since it's a closer match to how the graphs are originally organized.

**Comment by Hamms** [comment]
@sfilman thoughts on #2?

**Comment by Hamms** [comment]
How do you feel about something like this

image
image

File: dashboard/public/frequency/frequency.js [hunk 1]
```diff
@@ -362,104 +469,110 @@ BarGraph.prototype.reorder = function () {
...
 
-  // cache the english -> user mapping
-  var substMap = this.english_data.reduce(function (map, d, i) {
-    map[d.letter] = this.user_data[i];
+  // cache the message -> substitution mapping
+  // Note that we don't use this.getSubstitutionMap here, as that
+  // ignores unlocked substitutions
+  var substMap = this.message_data.reduce(function (map, d, i) {
+      map[d.letter] = this.substitution_data[i];
     return map;
```
Comment (joshlory) [inline dashboard/public/frequency/frequency.js]: Nit: indentation.

File: dashboard/public/frequency/frequency.js [hunk 2]
```diff
@@ -699,40 +903,66 @@ BarGraph.prototype.shift = function (amt) {
...
-    unlocked_indexes[m] = y;
-    unlocked_indexes[i] = x;
+/**
+ * Sorts this.substitution_data by frequency
+ */
+BarGraph.prototype.sortSubstitutions = function () {
+  if (!this.confirmSortOkay()) {
+    return;
+  }
+  this.substitutions_data = this.substitution_data.sort(BarGraph.frequencySort);
```
Comment (joshlory) [inline dashboard/public/frequency/frequency.js]: Nit: not used (one more below).

File: dashboard/public/frequency/frequency.js [hunk 3]
```diff
@@ -469,39 +582,56 @@ BarGraph.prototype.createDragBehavior = function () {
...
-      if (index == i) return user_data[j];
-      else if (index == j) return user_data[i];
-      else return d;
-    });
+      letter_outline.attr({
+        "visibility": "visible",
+        "transform": "translate(" + (this.messageLetterScale(this.substitution_data[i].letter) + 0.5) + "," + (this.getHeight()+54) + ")"
+      });
+      bar_outline.attr({
+        "visibility": "hidden",
```
Comment (joshlory) [inline dashboard/public/frequency/frequency.js]: Nit: trailing comma.

File: dashboard/public/frequency/frequency.js [hunk 4]
```diff
@@ -568,43 +728,41 @@ BarGraph.prototype.buildSVG = function () {
...
+    .call(this.yTopAxis)
+    .append("text")
     .attr({
-      "class": "fa lockicon",
+      "transform": "rotate(-90)",
+      "y": 6,
       "dy": ".71em",
-      "y": 60
-    });
+      "class": "graphlabel",
```
Comment (joshlory) [inline dashboard/public/frequency/frequency.js]: Nit: trailing comma.

File: dashboard/public/frequency/frequency.js [hunk 5]
```diff
@@ -622,74 +780,120 @@ BarGraph.prototype.buildSVG = function () {
...
   this.refreshDragBehavior();
 
 };
 
 /**
- * Reorders this.user_data to represent a shifted alphabetic ordering.
- * If the user has locked any letters, confirms with the user then
- * clears the locks.
- * @param {number} amt - amount to shift by
+ * Unlocks all substutions
```
Comment (joshlory) [inline dashboard/public/frequency/frequency.js]: Nit: spelling (one more below).

**Comment by joshlory** [comment]
LGTM.  Lucky PR #4000! :trophy:

Feel free to ignore any or all nits.

**Comment by sfilman** [comment]
Chatted with @Hamms we're going to go with Caesar substitution and Random substitution to match the Lesson Plan & Activity Guide

**Pull request merged**